processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java

```diff
@@ -187,11 +199,75 @@ public RowSignature resultArraySignature(final ScanQuery query)
     }
   }
 
+  @Override
+  public boolean canFetchResultsAsFrames()
+  {
+    return true;
+  }
+
+  @Override
+  public Sequence<FrameSignaturePair> resultsAsFrames(
+      final ScanQuery query,
+      final Sequence<ScanResultValue> resultSequence
+  )
+  {
+    return resultSequence.map(
+        result -> {
+          final List rows = (List) result.getEvents();
+          final Function<?, Object[]> mapper = getResultFormatMapper(query);
+          final Iterable<Object[]> formattedRows = Iterables.transform(rows, (Function) mapper);
+
+          RowBasedCursor cursor = IterableRowsCursorHelper.getCursorFromIterable(
+              formattedRows,
+              result.getRowSignature()
+          );
+
+          FrameWriterFactory frameWriterFactory = FrameWriters.makeFrameWriterFactory(
+              FrameType.ROW_BASED,
+              new SingleMemoryAllocatorFactory(HeapMemoryAllocator.unlimited()),
+              result.getRowSignature(),
+              new ArrayList<>(),
+              true
+          );
+
+          Frame frame;
+
+          try (final FrameWriter frameWriter = frameWriterFactory.newFrameWriter(cursor.getColumnSelectorFactory())) {
+            while (!cursor.isDone()) {
+              if (!frameWriter.addSelection()) {
+                throw new FrameRowTooLargeException(frameWriterFactory.allocatorCapacity());
+              }
+
+              cursor.advance();
+            }
+
+            frame = Frame.wrap(frameWriter.toByteArray());
+          }
+
+          return new FrameSignaturePair(frame, result.getRowSignature());
```

Comment:
We are making one frame per result sequence and each result sequence represents one segment. This does not seem very scalable. 
Lets leave a note here so that we can get back to this in a future PR.

Reply:
Since we can have different row sigs per segment, what we can do is only start a new frame when the row signature is different. This will reduce the number of frames by a lot.